### ISSUE TITLE
Command "fg 1" doesn't work

### ISSUE THREAD

**Issue by BarbzYHOOL** [fish-shell/fish-shell#5019]
fish, version 2.7.1
2.7.1

Run two programs and ctrl +z both

Now you have two jobs in the background.

When I type "fg 1" or "fg 2" it doesn't work, but "fg %1" works

How to set it to use "fg 1" instead? like in bash

**Comment by faho** [comment]
A bunch of our job commands currently expect _PID_s, with the "%num" stuff then expanding to one.

That honestly seems kind of useless to me - why not just take a job id directly? The main problem I can see is that it would make `disown (jobs -p)` (which is an example mentioned in the documentation) fail, so this would be a breaking change without some trickery.

**Comment by BarbzYHOOL** [comment]
I'm not sure if you're saying that using "fg 1" would be useless but it's just that it's more practical to use and especially when coming from bash (or when switching with bash and fish)

But any way to make an alias especially for that case (for my own egoistical use? :D)?

**Comment by faho** [comment]
>I'm not sure if you're saying that using "fg 1" would be useless

Nooo, other way around! `fg 1`, where `1` is a _JOB ID_ is useful. `fg 1`, where `1` is a PID is annoying. That's because, to get the PID, you need to necessarily run something like `jobs -p` or `%` expansion (which we've actually removed).

The "PID" bit also makes you think that fish can operate on arbitrary processes (a PID of 1 would be your init process), which it can't - it can only operate on its own jobs.

**Comment by mqudsi** [comment]
This is a bash anachronism, and the recent changes to `%` overall where aimed at changing the behavior to be more consistent with bash where job-related commands take `%`-prefixed job ids. I prefer this way because if there are functions or builtins that ever gain the ability to operate on both PIDs and job ids, this would allow for a clear distinction between the two.

In general a process-related task taking a number takes a process id and not a job id. It doesn't take all that much effort to type in `%` and *usually* you're only `fg`ing the last task anyway..

**Comment by floam** [comment]
Agree on keeping % for these functions. We can make this easier to understand by improving the completions. And e.g. the `fg` documentation must explain this.

**Comment by faho** [comment]
>I prefer this way because if there are functions or builtins that ever gain the ability to operate on both PIDs and job ids, this would allow for a clear distinction between the two.

Which of our builtins could ever gain that ability?

Here's a list of our builtins that operate on jobs:

- bg
- disown
- fg
- jobs
- wait

jobs, wait and disown fundamentally _can't_ - we can only `wait()` on our children, and jobs and disown are explicitly about them.

And I think pgroups throw a wrench into fg and bg? Even if they don't, how do we return to the foreground after a non-child we fg'd ends? We don't get SIGCHLD.

And even if one of them does, could we then not also start offering a thing to get the PID then, instead of now always requiring some weird `%` expression or `jobs -p` incarnation?

**Comment by lx-931203** [comment]
It's much easier and more comfortable to type `fg 2` than `fg %2`, for example.

There is no any possible excuse here - shell must be comfortable to use, at least - less character to type.

**Comment by faho** [comment]
>There is no any possible excuse here

@lx-931203: The "excuse" here is backwards compatibility, and compatibility with bash. While we don't require compatibility, in practice it can be confusing if you change something that used to be compatible, so we try not to do it unless there's a good reason.

In this case there might be, but we'd have to figure out how this works together with `jobs -p`, and nobody has done that yet.

>shell must be comfortable to use, at least - less character to type.

It has *never* been fish's goal to be the shell that needs the least amount of characters. In fact when it comes to code golf, I'm pretty sure zsh is much more successful.

It's also just one additional character whenever you need this, so there's no pressing need to figure this out, and it fits in with the broader "job management" thing that requires a bunch more work, but also just a deep understanding of what it should actually *do*. Because right now, I can't really see too much of a use for background jobs - you can just open another terminal.

**Comment by lx-931203** [comment]
> Because right now, I can't really see too much of a use for background jobs - you can just open another terminal.

Background jobs are super useful - server or any other stuff works in background and print to main tty only errors.

**Comment by ridiculousfish** [comment]
For my own record this tracks the distinction between job IDs (currently requiring %) and process ID (without %). The default is probably wrong but there are compatibility concerns with changing it.

My feeling is that this can probably be fixed to interpret a raw number as a job ID. In practice the number of background jobs is typically 0-2, and these are not likely to collide with real pids. But I'm not completely sold.

If/when fish supports background functions/blocks, this will become more pertinent because these background "processes" will be implemented via threads, not kernel processes, and so will not have a pid.

**Comment by faho** [comment]
>In practice the number of background jobs is typically 0-2, and these are not likely to collide with real pids. 

It is in fact very likely once you're running in a container. We've had issues with fish killing pgroup 2 (i.e. the "INVALID_PID" value).

Allowing both without distinguishing them is a non-starter.

**Comment by ridiculousfish** [comment]
> It is in fact very likely once you're running in a container. We've had issues with fish killing pgroup 2 (i.e. the "INVALID_PID" value).

Oh nice, today I learned!

**Comment by horst5000** [comment]
I must admit that the job handling is very weird in fish, since the job id seems to be pretty random. Quick example, where I start different tools in the prescribed order and send them to background (`ccmake, nano, man gcc, vim, vi` where vi is an alias to `nvim -O`). The output of jobs is extremely unexpected, since `man gcc` will receive jobid 5 instead of 3. `vim` started as fourth tool will receive 4. The `nvim` alias started last will receive 7.

```
jobs
Job     Group   CPU     State   Command
7       30462   0%      stopped nvim  $argv
4       30393   0%      stopped vim
5       30348   0%      stopped command man $argv
2       30308   0%      stopped nano
1       30273   0%      stopped ccmake .
```
Or simply starting `man gcc` and then `vim` results in job 1 and to not even being present
```
Job     Group   CPU     State   Command
4       1716    0%      stopped nvim  $argv
3       1660    0%      stopped command man $argv
```
In short, the job ids are not consecutive, which is weird, but even worse they are not even monotonically increasing. This effectively makes the use of `fg/bg %id` trial and error and significantly less useful compared to most other shells.

**Comment by sekoenig** [comment]
I had the same issue as discussed in this thread (wanting to be able to type `fg 1` instead of `fg %1`), and had the following naive idea to solve it:
```fish
function fg -a id                                                                    
  builtin fg %$id                                                                    
end  
```
Alas, it doesn't work:
```
fg: Argument '%1' is not a valid integer
```
Therefore, I have a question: how can I manually call process expansion for based on the value of `$id`?

**Comment by sekoenig** [comment]
Okay, found it in #5286.  The following works:
```fish
function fg -a id
  builtin fg (__fish_expand_pid_args %$id)
end
```

**Comment by marcusatiliusregulus** [comment]
This definitely needs to be documented; i was confused too and could not find anything in the docs. Not everyone (me for example) knows about the %-prefix for example from bash (thats how i figured it out in the end; should have looked here first...)

I would change the documentation like this for example (diff on bb0d4ed)

```
11c11
<     bg [PID...]
---
>     bg [ PID | JOBID ... ]
18a19
> You can also name a job by it's JOBID (the "Job" column in ``jobs``' output) prefixed with '%'.
32a34,35
>
> ``bg %1`` will resume the the job with the jobid 1 in the background.
```

but i dont know if pids and jobids can be mixed and other specificities and currently do not have the time to investigate...

**Comment by zanchey** [comment]
I've added some language to the `bg`, `fg`, and `jobs` documentation to explain this.

**Comment by faho** [comment]
@wllenyj What is that supposed to do?

Is that supposed to be a synonym for the last job?

**Comment by faho** [comment]
In that case, we're not inclined to add hard-to-understand shortcuts like that, no.

**Comment by wllenyj** [comment]
https://www.gnu.org/software/bash/manual/html_node/Job-Control-Basics.html

> The previous job may be referenced using ‘%-’. If there is only a single job, ‘%+’ and ‘%-’ can both be used to refer to that job. In output pertaining to jobs (e.g., the output of the jobs command), the current job is always flagged with a ‘+’, and the previous job with a ‘-’.

**Comment by krobelus** [comment]
This is neat, I think it makes sense to add `fg -`.

**Comment by joseluis** [comment]
Starting to use fish coming from bash this was something I truly missed. Thank you for the script, I've named it `fg_` and now I feel complete again.

**Comment by wllenyj** [comment]
@joseluis A spelling error.
`set --local real_command (ps -q $pid -o cmd=)` -> `ps -p`
